save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/controllers/DownloadProjectController.kt

```diff
@@ -126,29 +117,114 @@ class DownloadProjectController(
+        // search or create new test suites source by content
+        val testSuitesSourceMono = preprocessorToBackendBridge.getTestSuitesSource(
+            organizationName,
+            gitUrl,
+            testRootPath,
+            branch
+        )
+        val testSuitesSourceWithVersion = version
+            ?.let { testSuitesSourceMono.fetchSpecificTestSuites(it) }
+            ?: testSuitesSourceMono.fetchLatestTestSuites()
+        return testSuitesSourceWithVersion.flatMap { (testSuitesSource, version) ->
+            triggerTests(testSuitesSource, version, requestBase)
+        }
+    }
+
+    private fun Mono<TestSuitesSource>.fetchLatestTestSuites() = this
+        // fetch new test suites if required
+        .flatMap { testSuitesSource ->
+            testSuitesPreprocessorController.fetch(testSuitesSource.toDto())
+                .map { testSuitesSource }
+        }
+        // take latest version from backend
+        .zipWhen { preprocessorToBackendBridge.getTestSuitesLatestVersion(it.organization.name, it.name) }
+
+    private fun Mono<TestSuitesSource>.fetchSpecificTestSuites(providedVersion: String) = this
+        // fetch new test suites if required
+        .flatMap { testSuitesSource ->
+            testSuitesPreprocessorController.fetch(testSuitesSource.toDto(), providedVersion)
+                .map { testSuitesSource }
+        }
+        .zipWith(Mono.just(providedVersion))
+
+    private fun fetchAndTriggerSpecificTests(
+        organizationName: String,
+        gitUrl: String,
+        testRootPath: String,
+        branch: String,
+        providedVersion: String,
+        requestBase: ExecutionRequestBase,
+    ): Mono<StatusResponse> {
+        // search or create new test suites source by content
+        val testSuitesSourceMono = preprocessorToBackendBridge.getTestSuitesSource(
+            organizationName,
+            gitUrl,
+            testRootPath,
+            branch
+        )
+        return testSuitesSourceMono
+            // fetch new test suites if required
+            .flatMap { testSuitesSource ->
+                testSuitesPreprocessorController.fetch(testSuitesSource.toDto(), providedVersion)
+                    .map { testSuitesSource }
+            }
+            .flatMap { testSuitesSource ->
+                triggerTests(testSuitesSource, providedVersion, requestBase)
+            }
+    }
+
+    private fun triggerTests(
+        testSuitesSource: TestSuitesSource,
+        version: String,
+        requestBase: ExecutionRequestBase,
+    ) = preprocessorToBackendBridge.getTestSuites(
+        testSuitesSource.organization.name,
+        testSuitesSource.name,
+        version
+    )
+        .flatMapMany { Flux.fromIterable(it) }
+        .filter(requestBase.getTestSuiteFilter())
+        .collectList()
+        .flatMap { testSuites ->
+            updateExecution(
+                requestBase.project,
+                testSuites.map { it.version }
```

Comment:
Current logic doesn't expect different version for test suites on a single execution run

Reply:
added validation